### PR TITLE
Add Management API for markets & blockchains & wallets (closes #786, closes #1602)

### LINKED ISSUE #786: [Management API v1] Operations about markets

### LINKED ISSUE #1602: [Management API v1] Operations about blockchains and wallets

### PR DESCRIPTION
**Pull request by mnaichuk**

### CODE REVIEW

File: spec/api/management_api_v1/markets_spec.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        write_markets: { permitted_signers: %i[alex jeff], mandatory_signers: %i[alex] }
+    }
+  end
+
+  describe 'List markets' do
+    def request
+      post_json '/management_api/v1/markets', multisig_jwt_management_api_v1({ data: data }, *signers)
+    end
+
+    let(:data) { {} }
```
Comment (kovetskiy) [inline spec/api/management_api_v1/markets_spec.rb]: seems like unused :data variable

**Comment by ysv** [comment]
@Naichuk fix specs

**Comment by mnaichuk** [comment]
@ysv Done

**Comment by ysv** [comment]
@mnaichuk add new scopes to management_api_v1.yml.erb

File: app/api/management_api_v1/blockchains.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    params do
+      requires :key, type: String, desc: 'Blockchain key'
+      requires :name, type: String, desc: 'Blockchain name'
+      requires :client, type: String, desc: 'Blockchain client'
+      requires :server, type: String, desc: 'Blockchain server'
+      requires :height, type: Integer, desc: 'Blockchain height'
+      requires :explorer_address, type: String, desc: 'Blockchain explorer address'
+      requires :explorer_transaction, type: String, desc: 'Blockchain explorer transaction'
+      optional :min_confirmations, type: Integer, default: 6, desc: 'Minimum confirmations from network'
+      optional :status, type: String, default: 'active', desc: 'Blockchain status'
```
Comment (ysv) [inline app/api/management_api_v1/blockchains.rb]: `Blockchain status active or disabled`

File: app/api/management_api_v1/blockchains.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      present Blockchain.all, with: ManagementAPIv1::Entities::Blockchain
+      status 200
+    end
+
+    desc 'Return blockchain by key.' do
+      @settings[:scope] = :read_blockchains
+      success ManagementAPIv1::Entities::Blockchain
+    end
+    params do
+      requires :key, type: String, desc: 'Blockchain key'
```
Comment (ysv) [inline app/api/management_api_v1/blockchains.rb]: validate values

File: app/api/management_api_v1/blockchains.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    end
+    post 'blockchains/get' do
+      present Blockchain.find_by(key: params[:key]), with: ManagementAPIv1::Entities::Blockchain
```
Comment (ysv) [inline app/api/management_api_v1/blockchains.rb]: use `find_by!`

File: app/api/management_api_v1/blockchains.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      optional :client, type: String, desc: 'Blockchain client'
+      optional :server, type: String, desc: 'Blockchain server'
+      optional :height, type: Integer, desc: 'Blockchain height'
+      optional :explorer_address, type: String, desc: 'Blockchain explorer address'
+      optional :explorer_transaction, type: String, desc: 'Blockchain explorer transaction'
+      optional :min_confirmations, type: Integer, default: 6, desc: 'Minimum confirmations from network'
+      optional :status, type: String, default: 'active', desc: 'Blockchain status'
+    end
+    put 'blockchains/update' do
+      blockchain = Blockchain.find_by(key: params[:key])
```
Comment (ysv) [inline app/api/management_api_v1/blockchains.rb]: use `find_by!`

File: app/api/management_api_v1/markets.rb [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      present Market.find(params[:id]), with: ManagementAPIv1::Entities::Market
+      status 200
+    end
+
+    desc 'Create new market' do
+      @settings[:scope] = :write_markets
+      success ManagementAPIv1::Entities::Market
+    end
+    params do
+      optional :id, type: String, desc: 'Market id.'
```
Comment (ysv) [inline app/api/management_api_v1/markets.rb]: remove id

File: app/api/management_api_v1/markets.rb [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    end
+
+    desc 'Update exist market' do
+      @settings[:scope] = :write_markets
+      success ManagementAPIv1::Entities::Market
+    end
+    params do
+      requires :id, type: String, desc: 'Market id.'
+      optional :ask_unit, type: String, desc: 'Market ask unit'
+      optional :bid_unit, type: String, desc: 'Market bid unit'
```
Comment (ysv) [inline app/api/management_api_v1/markets.rb]: There are some read only fields in Market please check and remove

File: app/api/management_api_v1/markets.rb [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      requires :bid_fee, type: BigDecimal, desc: 'Market bid fee'
+      optional :max_bid, type: BigDecimal, desc: 'Market max bid'
+      optional :min_ask, type: BigDecimal, default: 0, desc: 'Market min ask'
+      requires :ask_precision, type: Integer, desc: 'Market ask precision'
+      requires :bid_precision, type: Integer, desc: 'Market bid precision'
+      optional :position, type: Integer, desc: 'Market position'
+      optional :enabled, type: Boolean, default: true, desc: 'Market status'
+    end
+    post 'markets/new' do
+      market = Market.new
```
Comment (ysv) [inline app/api/management_api_v1/markets.rb]: `market = Market.new(params)`
remove `market.assign_attributes`

File: spec/api/management_api_v1/markets_spec.rb [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  end
+
+  describe 'Update market' do
+    def request
+      put_json '/management_api/v1/markets/update', multisig_jwt_management_api_v1({ data: data }, *signers)
+    end
+
+    let(:data) { { id: 'btcusd', ask_fee: 0.1 } }
+    let(:signers) { %i[alex jeff] }
+    let(:expected_response) { Market.find(data[:id]).update('ask_fee': 0.1)}
```
Comment (ysv) [inline spec/api/management_api_v1/markets_spec.rb]: remove this let

**Comment by ysv** [comment]
@mnaichuk add entities specs

**Review by ysv — CHANGES_REQUESTED** [review]

File: spec/api/management_api_v1/blockchains_spec.rb [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    def request
+      put_json '/management_api/v1/blockchains/update', multisig_jwt_management_api_v1({ data: data }, *signers)
+    end
+    let(:data) do
+      { key:                'xrp-testnet',
+        min_confirmations:  5,
+        height:             11111111}
+    end
+    let(:signers) { %i[alex jeff] }
+    let(:do) { Blockchain.find_by!(key: data[:key]).update(min_confirmations: 5) }
```
Comment (ysv) [inline spec/api/management_api_v1/blockchains_spec.rb]: This one is not used I think

File: spec/api/management_api_v1/markets_spec.rb [hunk 11]
```diff
@@ -0,0 +1,119 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+describe ManagementAPIv1::Deposits, type: :request do
```
Comment (ysv) [inline spec/api/management_api_v1/markets_spec.rb]: `ManagementAPIv1::Deposits`

File: spec/api/management_api_v1/blockchains_spec.rb [hunk 12]
```diff
@@ -0,0 +1,120 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+describe ManagementAPIv1::Deposits, type: :request do
```
Comment (ysv) [inline spec/api/management_api_v1/blockchains_spec.rb]: ` ManagementAPIv1::Deposits`

File: spec/api/management_api_v1/markets_spec.rb [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        bid_unit:       'btc',
+        ask_fee:        0.001,
+        bid_fee:        0.001,
+        ask_precision:  8,
+        bid_precision:  8,
+        enabled:        true }
+    end
+    let(:signers) { %i[alex jeff] }
+
+    it 'creates new market with enabled state' do
```
Comment (ysv) [inline spec/api/management_api_v1/markets_spec.rb]: it 'creates new market with expected attributes'

File: spec/api/management_api_v1/blockchains_spec.rb [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+       server:                'http://user:password@127.0.0.1:5005',
+       height:                11773777,
+       explorer_address:      'https://bithomp.com/explorer/\#{address}',
+       explorer_transaction:  'https://bithomp.com/explorer/\#{txid}',
+       min_confirmations:     6,
+       status:                'active' }
+    end
+    let(:signers) { %i[alex jeff] }
+
+    it 'creates new blockchain with state active' do
```
Comment (ysv) [inline spec/api/management_api_v1/blockchains_spec.rb]: it 'creates new blockchain with expected attributes'

File: spec/api/management_api_v1/blockchains_spec.rb [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    end
+    let(:data) do
+      { key:                'xrp-testnet',
+        min_confirmations:  5,
+        height:             11111111}
+    end
+    let(:signers) { %i[alex jeff] }
+    let(:do) { Blockchain.find_by!(key: data[:key]).update(min_confirmations: 5) }
+
+    it 'updates blockchain' do
```
Comment (ysv) [inline spec/api/management_api_v1/blockchains_spec.rb:111]: can you also add test that it doesn't allow to update read-only attributes. Example client

File: spec/api/management_api_v1/markets_spec.rb [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+      put_json '/management_api/v1/markets/update', multisig_jwt_management_api_v1({ data: data }, *signers)
+    end
+
+    let :data do
+      { id: 'btcusd',
+        ask_fee: 0.1 }
+    end
+    let(:signers) { %i[alex jeff] }
+
+    it 'updates market' do
```
Comment (ysv) [inline spec/api/management_api_v1/markets_spec.rb:110]: can you also add test that it doesn't allow to update read-only attributes. Example bid_unit

**Review by ysv — CHANGES_REQUESTED** [review]

**Comment by hardywu** [comment]
is this feature gonna implemented?

**Pull request closed without merging**